### DISCUSSION TITLE
TIP - Metadata Encoding Standard v1

### DISCUSSION THREAD

**Discussion by Tuditi** [iotaledger/tips#131] [Ideas]
# TIP - Metadata Encoding Standard v1 (MES-1)

# Preamble

---

tip: ?
title: Metadata Encoding Standard v1 (MES-1)
description: 
author: Nicole O'Brien (@nicole-obrien) <cheerfulnicole9@gmail.com>, David De Troch (@tuditi) <david.detroch@pm.me>, Matthew Maxwell (@maxwellmattryan) <maxwellmattryan@gmail.com>
status: Draft
type: Process
created: 2023-05-12

---

# Abstract

The TIP describes the Metadata Encoding Standard v1 (MES-1), a standard for encoding and decoding data in the metadata field of an output. The lack of a standard for metadata encoding/decoding poses a problem for applications that interpret metadata of an output/transaction across the ecosystem. The specification outlines a layered approach to encoding, with each encoding level being a layer of the onion. Level 1 determines the encoding of the data type, with backwards compatibility for existing metadata that uses JSON. Level 2 defines the format of the data. Further levels of encoding can be defined for specific standards. The specification also proposes a flexible system for defining the bytes that tell us how to decode the data, which allows to add future standards. The proposed encoding uses the first bit as a flag and the remaining bits to denote the encoding type.

# Motivation

Currently, applications don't have a standard that helps them encode and decode the data fields of a Transaction or Output (e.g. the metadata field of an Alias). Each application is free to define data objects as they see fit, even if the data format already adheres to an RFC standard (e.g. JSON, UTF-8, CBOR, ...). This poses a problem for applications that interpret metadata across the ecosystem. We think it's beneficial to introduce a TIP that standardizes encoding metadata.

## Encoding / Decoding

This TIP defines several levels of encoding, much like the layers of an onion. Each layer has to be decoded in order to understand the full message. For simplicity we will order the layers in the order required to decode and read the data at the center of the onion.

This TIP refers to the State Metadata field of an Alias Output and the Metadata Feature as "metadata" throughout the document. [TIP-18](https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md) defines the byte format of the metadata on a lower level.

### Layer 1 - Encoding

Layer 1 is used to determine how we should decode the metadata e.g. should we treat the data as a byte array or a UTF8 string or something else entirely.

Initial sensible layer 1 encodings can be defined as:

| Value | Hex | Encoding         | Check Next Encoding    | Type      | Comment                                       |
| ----- | --- | ---------------- | --- | --------- | --------------------------------------------- |
| 0     | 00  | Byte Array       |  True   | Bytes     |                                               |
| 8     | 08  | UTF8             |  True   | Text      |                                               |
| 16    | 10  | UTF16            |  True   | Text      |                                               |
| 18    | 12  | GB18030          |  True   | Text      |                                               |
| 32    | 20  | UTF32            |  True  | Text      |                                               |
| 34    | 22  | UTF8/JSON/string |   False  | Text/JSON | Special case for JSON backwards compatibility |
| 49    | 31  | Bytes/CBOR             |   False  | Bytes     |                                               |
| 91    | 5B  | UTF8/JSON/array  |  False   | Text/JSON | Special case for JSON backwards compatibility |
| 123   | 7B  | UTF8/JSON/object |   False  | Text/JSON | Special case for JSON backwards compatibility |

### Layer 2 - Format

Layer 2 determines the format of the data that the application wants to read. This could represent an existing data or file format or a new custom proprietary one that can be linked in this TIP e.g. JSON, CBOR, XML.

##### Byte Array - 00

| Value | Hex | Encoding                | Check Next Byte    | Type        | Comment                                                         |
| ----- | --- | ----------------------- | --- | ----------- | --------------------------------------------------------------- |
| 0     | 00  | Bytes                   |  False   | Plain Bytes |                                                                 |
| 1     | 01  | ISC |   true  | Plain Bytes | Need to define the standards for smart contract metadata fields |
49 |	31 |	CBOR |	False |	Bytes	

##### Text Format - 08, 10, 12, 20

| Value | Hex | Encoding | Check Next Byte    | Type       | Comment |
| ----- | --- | -------- | --- | ---------- | ------- |
| 0     | 00  | String   |   False  | Plain Text |         |
| 1     | 01  | JSON     |   False  | Document   |         |

### Layer 3 - Standard

Layer 3 is an optional standard that can be defined to allow programs to understand the data. Decoding the standards will be specific to the data format chosen. E.g IRC30, defined in [TIP-30](https://github.com/iotaledger/firefly/pull/6991/files), contains the definition of how to store metadata of a native token and currently only contains a definition for JSON format but it could be extended to include a decoding for CBOR etc.

##### JSON

In JSON a standard can be defined using the key `standard` and an optional version defined using the key `version`. The [JSON scheme](#json-schema-for-layer-3-standards) at the end of this document provides information around its structure. Currently we have two defined standards for JSON formatted metadata:	

 Standard | Version | Link | 
| ----- | --- | -------- | 
| IRC27     | v1.0  | https://github.com/iotaledger/tips/blob/main/tips/TIP-0027/irc27.schema.json
| IRC30     |   | https://github.com/iotaledger/tips/blob/main/tips/TIP-0030/irc30.schema.json

##### ISC
| Value | Hex | Encoding             | Check Next Byte | Link                                              |
| ----- | --- | -------------------- | --------------- | ------------------------------------------------- |
| 0     | 00  | ISC state            | False           |                                               |
| 1     | 01  | ISC on chain request | False           | https://hackmd.io/SWpeUi_eTDKbnvK623vMmw#Metadata |

#### Combinations

We also have the ability to shorten encodings to fewer bytes by defining a specific byte for a combination of the above encodings. And in specific cases as we will describe later for backwards compatibility the the byte itself could form part of the data to be parsed.

### Encoding bytes

The metadata is represented by binary data, so bytes are a logical choice for markers or flags to decode the data. Therefore we introduce a system to define the bytes which tell us how to decode the data. 

#### Extendable bytes

The system is flexible so new standards can be added in the future. The proposed encoding uses the first bit as a flag and the remaining bits to denote the encoding type. If the flag bit is unset (flag bit = 0), only a single byte signals the format decoding type. Otherwise (flag bit = 1), the decoding also considers the next byte for decoding the format.

![](https://i.imgur.com/HXZ84PZ.jpg)

![](https://i.imgur.com/IyCaCFU.jpg)

This approach provides 128 single byte values to define standards for. An additional 32,760 double byte values ensure that the encoding is future proof. 

# Rationale

Currently, two application level metadata standards exist: the IRC27 and IRC30 metadata standards for NFTs and Foundries. They don't specify how they are encoded, but applications assume that the metadata field consists of hex encoded JSON objects. 

Soon we will introduce three more metadata standards. These relate to the anchoring of smart contract state and DID records and the interaction with the smart contract chains on layer 1. To our knowledge the two smart contract standards save space through proprietary encoding. (There is no need to store a key value pair / object if you know all the fields and can enforce the size and order).

These will not be the only metadata standards introduced in IOTA's lifetime, and applications should decode the metadata without using trial and error. Therefore, we propose a metadata encoding standard before adding any more functionality.

This design represents an efficient encoding that optimizes the space used on the Tangle, but allows future additions to the standard. The standard is backwards compatible with the basic JSON metadata to minimize changes for the TIPs concerning token foundries or NFTs.

# Backwards Compatibility

Currently there is no standard for encoding metadata, and each individual standard, such as IRC-27 and IRC-30 states how they encode the metadata field. By introducing this encoding standard, previous stored metadata may contradict the standard.

MES-1 covers the current known metadata standards on the Shimmer network. The only two published standards, to the best of our knowledge, are:
- IRC27
- IRC30

The analysis of these standards resulted in exceptions to the layered encoding approach while also minimizing the bytes used for more common standards. 

IRC27 and IRC30 metadata use UTF8/JSON encoding to represent objects in JSON. The JSON standard determines what the initial characters can be. MES-1 supports JSON objects, arrays and strings (i.e., objects are initialized by `{`, arrays by `[` and strings by `"`). The layer 1 encoding reserves the hex value of these starting characters to ensure maximum backwards compatibility.

* `{` is represented by `123` in UTF8, therefore the hex code reserved is `7B`
* `[` is represented by `91` in UTF8, therefore the hex code reserved is `5B`
* `"` is represented by `34` in UTF8, therefore the hex code reserved is `22`

# Examples

#### Metadata:
`[01010100, 00000110, 01010111, 10000110, 00010110, 11010111, 00000110, 11000110, 01010010, 11100110, 00110110, 11110110, 11010010, 00100010, 11000000, 10100010, 00000010, 00001111]`

#### HEFS

| Layer        | Encoding |
| ------------ | -------- |
| 0 - Hex Flag | 0x       |
| 1 - Encoding | 00       |
| 2 - Format   | 00       |

#### Encoded Metadata
`0x00005406578616d706c652e636f6d222c0a2020f`

#### Metadata

``'Hello World'``

#### HEFS

| Layer        | Encoding |
| ------------ | -------- |
| 0 - Hex Flag | 0x       |
| 1 - Encoding | 08       |
| 2 - Format   | 00       |

#### Encoded Metadata

`0x080048656c6c6f20576f726c64`

#### Metadata

```json
{
  "id": "12345",
  "name": "John Doe",
  "age": 30,
  "email": "johndoe@example.com",
  "address": {
    "street": "123 Main St",
    "city": "Anytown",
    "state": "CA",
    "zip": "12345"
  }
}
```

#### HEFS

| Layer        | Encoding |
| ------------ | -------- |
| 0 - Hex Flag | 0x       |
| 1/2 - Encoding | 7B       |

#### Encoded Metadata

Results in the following metadata:

`0x7b0a2020226964223a20223132333435222c0a2020226e616d65223a20224a6f686e20446f65222c0a202022616765223a2033302c0a202022656d61696c223a20226a6f686e646f65406578616d706c652e636f6d222c0a20202261646472657373223a207b0a2020202022737472656574223a2022313233204d61696e205374222c0a202020202263697479223a2022416e79746f776e222c0a20202020227374617465223a20224341222c0a20202020227a6970223a20223132333435220a20207d0a7d`

#### Metadata

```json
{
  "standard": "IRC30",
  "name": "Metadata Token",
  "symbol": "MES1",
  "decimals": 0
}
```

#### HEFS

| Layer        | Encoding |
| ------------ | -------- |
| 0 - Hex Flag | 0x       |
| 1/2 - Encoding | 7B       |
| 3 - Standard | IRC30       |

#### Encoded Metadata

`0x7b227374616e64617264223a224952433330222c226e616d65223a224d6574616461746120546f6b656e222c2273796d626f6c223a224d455331222c22646563696d616c73223a307d`

#### Metadata

```json
{
  "senderContract": 0,
  "targetContract": 1011572226,
  "entryPoint": 603251617,
  "params": {
    "Items": [
      {
        "key": "0x61",
        "value": "0x03cbcd6d8659ed1998a452335ae53904dc0af1c99b"
      }
    ]
  },
  "allowance": {
    "baseTokens": 25000000,
    "nativeTokens": [],
    "nfts": []
  },
  "gasBudget": 500000
}
```

#### HEFS

| Layer        | Encoding |
| ------------ | -------- |
| 0 - Hex Flag | 0x       |
| 1 - Encoding | 00       |
| 2 - Format   | 01       |
| 3 - Standard | 01       |

#### Encoded Metadata

`0x00010100000000025e4b3ca1e3f42320a1070000000000010000000100611500000003cbcd6d8659ed1998a452335ae53904dc0af1c99b0040787d0100000000020000000000`

### [JSON Schema for Layer-3 standards](#json-schema-for-layer-3-standards)

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://github.com/iotaledger/tips/main/tips/TIP-0038/irc38.schema.json",
  "title": "JSON Schema to define layer-3 JSON standards",
  "description": "JSON Schema to define layer-3 JSON standards",
  "type": "object",
  "properties": {
    "standard": {
      "type": "string",
      "description": "identifier for the metadata standard used",
    },
    "version": {
      "type": "string",
      "description": "version of the standard used",
      "pattern": "^v\\d+.\\d+$"
    }
  },
  "required": [ "standard" ]
}
```
# Questions

1. How can we partition the metadata field to store more than one piece of encoded data potentially with different hex encodings?
2. Do we want to define a layer 1 encoding for ISC smart contract requests to save 1 byte?
3. Do we want to define a layer 1 encoding for ISC state to save 1 byte
4. What is the governance/process for updating the Metadata encoding standards?
5. Can we provide an example repository for encoding and decoding.
6. If no encoding is detected, should we treat the metadata as a stringified JSON by default or omit the handling from this standard?
7. Is CBOR encoding / standard defined correctly?
8. How can we accommodate identity standards in this?

**Comment by nicole-obrien** [comment]
Thanks for posting this @Tuditi, I am looking forward to peoples thoughts and suggestions 😄

**Comment by eike-hass** [comment]
> How can we partition the metadata field to store more than one piece of encoded data potentially with different hex encodings?

I consider this critical to enable central outputs that can be used to represent/anchor multiple applications. 
An alternative to "partitioning" each meta data field would be allow multiple meta data fields, but that comes with more significant changes in the output layout.

**Reply by lzpap** [reply]
> How can we partition the metadata field to store more than one piece of encoded data potentially with different hex encodings?

Why don't you just set the metadata as JSON encoded string and put the different pieces of hex encoded data in the JSON level representation? You can define whatever application level JSON schema standard for DIDs or other use cases.

**Comment by eike-hass** [comment]
> If no encoding is detected, should we treat the metadata as a stringified json by default or omit the handling from this standard?

Could you elaborate who the "we" is in that scenario? I assume the protocol will not inspect / validate those fields. Is this concerning applications that would want to display the meta data without interpreting it like wallets and explorers? Interpreting the data would in many cases require the actual business logic of the specific application, I assume.

**Reply by Tuditi** [reply]
Yes, it concerns the higher level applications in general that interpret data according to this standard. The TIP doesn't make any assumptions on the actual business logic, which I also think is out-of-scope. 

An interesting question is whether there are security risks present if applications handle the metadata as 'trusted'. I think the answer to this question depends on the given application and it's important to disclaim that the fields are not validated.

**Comment by eike-hass** [comment]
> How can we accomodate identity standards in this?

From the perspective of the DID method (`did:iota`) there are no specific considerations. How to encode->store->retrieve->decode a DID document is defined in the specific method and varies dramatically between methods, so we would update our method to use the JSON Object encoding.

There are similar consideration to L2 SC chains regarding backward compatibility though. We can help implementers transition their documents through the library, but to not break existing documents (not under the implementers control) the library would need to support the current and the proposed encoding for the foreseeable future.

**Reply by Tuditi** [reply]
What is the current encoding for L2 SC chains?

**Comment by eike-hass** [comment]
Are there alternatives to HEX encoding as the meta data field are size constrained and the overhead might be prohibitiv for larger data chunks or co-existance/partitioning.

**Reply by Tuditi** [reply]
For now we didn't consider other encoding schemes. Of course an application that needs larger data chunks, can choose any other encoding, since the metadata field is not intended to be validated. Additionally, the standard can be extended by adding a specific encoding in the table for Layer 1. The data could he a hex encoded representation of a more efficient encoding scheme.

**Reply by eike-hass** [reply]
> The data could be a hex encoded representation of a more efficient encoding scheme.

Which then would it make less efficient again, no?

**Reply by Tuditi** [reply]
In the end the hex representation is just used to denote the bytes. So it's not that we encode the UTF-8 string of the hex encoding of the bytes. Maybe the following line regarding layer 0 should be changed from:

`The first layer of encoding/decoding defines that all metadata following this standard should be encoded as hexadecimal with a prefix of 0x.`

to:

`The standard expects that all bytes are represented in a hexadecimal format.`

What do you think?

**Reply by Thoralf-M** [reply]
> In the end the hex representation is just used to denote the bytes. 

I think this should be clarified a bit, there is also `Level 0 defines metadata following the standard as a hexadecimal with a prefix of 0x.`
To me it sounds a bit like the metadata is stored as hex encoded string

**Reply by Tuditi** [reply]
I edited it. Is it more clear like this? In the end level 0 doesn't know to much about what the metadata really represents. It just states that it's using hexadecimals to represent whatever else is following.

**Reply by Thoralf-M** [reply]
To me it sounds better now :+1:

**Comment by PhilippGackstatter** [comment]
> The TIP describes the Metadata Encoding Standard v1 (MES-1), a standard for encoding and decoding data in the metadata field of an output.

Can you clarify whether this refers to the [Metadata Feature](https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md#metadata-feature) that can be put in outputs or the `State Metadata` field of an [Alias Output](https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md#alias-output), or both? Thanks!

**Reply by Tuditi** [reply]
It refers primarily to the Metadata Feature, but I see no reason why this couldn't be extended to the `State Metadata` field.

**Comment by anistark** [comment]
Why `MES-1` ? Shouldn't all TIPs be referred as TIP-n only?
Might create confusing while referring later.

**Reply by Tuditi** [reply]
Metadata Encoding Standard is the name of the TIP & 1 is the version. Each TIP has to have a name that describes what it does. If this becomes accepted, the community can refer to it as TIP-38.

**Reply by lzpap** [reply]
Yeah it will be just called "TIP-XX: Metadata Encoding Standard".

**Comment by PhilippGackstatter** [comment]
Thanks for creating this TIP!  I have a few questions/comments.

> The TIP describes the Metadata Encoding Standard v1 (MES-1), a standard for encoding and decoding data in the metadata field of an output. 

> Currently, applications don't have a standard that helps them encode and decode the data fields of a Transaction or Output (e.g. the metadata field of an Alias). 

I think the TIP should be more specific and define whether this applies to the bytes in a _Metadata Feature_ or the `State Metadata` field of an Alias Output, or both.

> Layer 0 - Hex Flag
> The metadata is a hexadecimal string,

Both _Metadata Feature_ and the `State Metadata` field of an Alias Output define in TIP-18 how bytes are stored: as `(uint16)ByteArray` (defined in TIP-21). So I don't understand what is meant by "The metadata is a hexadecimal string". The explorer and node API show/return hexadecimal strings, but those are just implementation details of explorer and API and this TIP should build on top of TIP-18 instead, to be implementation-agnostic. So I think this TIP should not need to define a Layer 0 encoding, as that would effectively conflict with TIP-18, unless I'm misunderstanding something about this standard.

> In JSON a standard can be defined using the key standard and an optional version defined using the key version. Currently we have two defined standards for JSON formatted metadata:

If I understand correctly, this TIP would standardize this schema, so could you add the JSON schema that specifies what the types of the `standard` and `version` keys are, and what the format for other fields is? Are other top-level keys allowed, e.g.

```json
{
  "standard": "iota-did",
  "version": "0.7",
  "document": {
    "id": "did:iota:0xabc..."
  }
}
```

or does it have to be in a specific key, say `data`, e.g.

```json
{
  "standard": "iota-did",
  "version": "0.7",
  "data": {
    "document": {
      "id": "did:iota:0xabc..."
    }
  }
}
```

**Reply by Tuditi** [reply]
Hey Philipp,

Since the `Metadata Feature` and the `State Metadata` are very similar, I think both would benefit from this standard. I added a small section on this:

```
This TIP refers to both the Metadata Feature of a basic output and the State Metadata of an alias output as metadata throughout the document. [TIP-18](https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md) defines the byte format of the metadata on a lower level.
```

With regard to the layer 0 encoding. This has been a pain point raised before by @eike-hass & @Thoralf-M , so I removed the references to layer 0 encoding. We were confusing the binary representation with the binary data itself.

I'll add a JSON Schema for the layer 3 encoding of JSON Objects. New standards should be free to add their own top-level keys in my opinion.

Thanks a lot for your thorough review and let me know what you think of the changes!

**Reply by PhilippGackstatter** [reply]
Thanks for addressing those points, much appreciated!

I think the hex encoding is still part of the "Encoding bytes" section, I suppose that one has to be removed as well, then?

> This TIP refers to both the Metadata Feature of a basic output and the State Metadata of an alias output as metadata throughout the document. [TIP-18](https://github.com/iotaledger/tips/blob/main/tips/TIP-0018/tip-0018.md) defines the byte format of the metadata on a lower level.

I think it would suffice to say:

> This TIP refers to the `State Metadata` field of an Alias Output and the _Metadata Feature_ as "metadata" throughout the document.

The _Metadata Feature_ can be part of a Basic, NFT and Alias Outputs within the _Feature_ or (for the latter two) also the _Immutable Features_ section. The way it's written now would imply it only applies to the Feature in the Basic Output, which I assume is not what you want to restrict it to.

**Reply by Tuditi** [reply]
Yes, you're right. I updated the missing parts. :relaxed:

**Comment by lzpap** [comment]
> Do we want to define a layer 1 encoding for ISC smart contract requests to save 1 byte?

Hey @Tuditi , have you talked with the ISC team about this? While I get that it makes sense to be able to parse the content of a request, I believe ISC uses custom minimal encoding in requests in order to save space and therefore deposits as well. If you are an application (Firefly, Explorer, etc.) and want to display request data, you'll be able to infer whether a piece of bytes should be parsed as an ISC request or not by simply looking at the target `alias Id`: is this a chain I know/care about?

> Do we want to define a layer 1 encoding for ISC state to save 1 byte

IMO it doesn't make sense. ISC state in the `State Metadata` of an alias is essentially a block header with a cryptographic commitment. A hash of 32 bytes. Does it make sense to show to a user? Ever? Who else than the chain actually needs to read/modify this field?

> What is the governance/process for updating the Metadata encoding standards?

Through the usual [TIP process](https://github.com/iotaledger/tips/blob/main/tips/TIP-0001/tip-0001.md#tip-workflow). In order for such an application level TIP to be successful, you need to demonstrate that it has a working implementation (in this case open source lib for working with MES encoding/decoding, preferably in more languages), furthermore that there is application level support for it. A quote from TIP-1:

_"It is your responsibility as a TIP author to build community consensus around your idea. Involve as many people in the discussion as you can. Use social media platforms, Discord or Reddit to raise awareness of your idea."_

_"TIPs that have broad support are much more likely to make progress than those that don't receive any comments. Feel free to reach out to the TIP editors in particular to get help to identify stakeholders and obstacles."_

In this case I'd also ask for clarification with the ISC team and whether anything needs to change on their side with metadata handling. I'd also talk to NFT projects in the ecosystem just so they know that this is compatible with how things are today. You already involved DID people, that's great!

> Can we provide an example repository for encoding and decoding.

You should, and also libs to help others work with it. Since this is application level, a bare minimum would be a TypeScript SDK.

> If no encoding is detected, should we treat the metadata as a stringified JSON by default or omit the handling from this standard?

My understanding is that since you reserved the first possible characters of JSON on layer 1, it's probably not a JSON if you can't detect the encoding. I'd just treat it as plain simple bytes and omit the handling/further processing.

> How can we accommodate identity standards in this?

I'd say treat DID as JSON objects, why would you want to develop a special encoding just for one use case?

+1 bonus question: Why did you consider only CBOR? Did you look into other commonly used web3 encodings? How about [RLP](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/), [Borsh](https://borsh.io/) or [BCS](https://github.com/zefchain/bcs)?
